unitTests/ui/StarRatingTest.ts

```diff
@@ -0,0 +1,155 @@
+          for (let i = 0; i <= 5; i++) {
+            initStarRatingComponent(i.toString());
+
+            let starData = getActiveStars(test.cmp.element);
+            expect(starData.numStars).toBe(5);
+            expect(starData.numActiveStars).toBe(i);
+          }
+        });
+
+        it('should display no active stars when the rating provided is negative', () => {
+          const testRating = -Number.MAX_VALUE;
+          initStarRatingComponent(testRating.toString());
+
+          let starData = getActiveStars(test.cmp.element);
+          expect(starData.numStars).toBe(5);
+          expect(starData.numActiveStars).toBe(0);
+        });
+      });
+
+      describe('The label showing number of ratings', () => {
+        it('should not display when no value is provided for number of ratings', () => {
+          initStarRatingComponent((0).toString());
+          const children = test.cmp.element.getElementsByClassName(STAR_LABEL_CSS_CLASS);
+          expect(children.length).toBe(0);
+        });
+
+        it('should display a label with "No Ratings" shown when zero given', () => {
+          const testNumRatings = 0;
+          initStarRatingComponent((0).toString(), testNumRatings.toString());
+
+          const testLabel = test.cmp.element.lastElementChild;
+          expect(testLabel.className).toEqual(STAR_LABEL_CSS_CLASS);
+          expect(testLabel.textContent).toEqual(l('No Ratings'));
+        });
+
+        it('should display a label with "No Ratings" shown when a negative number is given', () => {
+          const testNumRatings = -Number.MAX_VALUE;
+          initStarRatingComponent((0).toString(), testNumRatings.toString());
+
+          const testLabel = test.cmp.element.lastElementChild;
+          expect(testLabel.className).toEqual(STAR_LABEL_CSS_CLASS);
+          expect(testLabel.textContent).toEqual(l('No Ratings'));
+        });
+
+        it('should display the number when they are provided', () => {
+          const testNumRatings = Number.MAX_VALUE;
+          initStarRatingComponent((0).toString(), testNumRatings.toString());
+
+          const testLabel = test.cmp.element.lastElementChild;
+          expect(testLabel.className).toEqual(STAR_LABEL_CSS_CLASS);
+          expect(testLabel.textContent).toEqual(`(${testNumRatings})`);
+        });
+      });
+
+      describe('When a different scale is provided', () => {
+        it('should display five stars with a number of active stars equal to new proportions', () => {
+          const newScale = 10;
+          for (let i = 0; i <= newScale; i++) {
+            initStarRatingComponent(i.toString(), undefined, newScale);
+
+            let starData = getActiveStars(test.cmp.element);
+            expect(starData.numStars).toBe(5);
+            expect(starData.numActiveStars).toBe(Math.floor(i / 2));
+          }
+        });
+
+        describe('If the scale is smaller than the rating', () => {
```

Comment:
I would reserve using `describe` blocks for when you have 2 or more tests in the block. Otherwise, I prefer to combine the `describe` description with the `it` description to reduce the nesting.

Reply:
Does nesting affect performance?

Reply:
I am not sure if jasmine processes the tests differently, but the main philosophy behind my comment is we want to reduce the code in all places to it's simplest form, but no simpler.

In this case, the `describe` block is not being used for it's main function, which is to group similar tests and testing utilities together. This gives us an opportunity to simplify.

Reply:
All right got it. Thanks